src/main/java/org/assertj/core/internal/Urls.java

```diff
@@ -48,6 +47,15 @@ public static Urls instance() {
 
   Urls() {}
 
+  static String extractNonQueryPart(URL url) {
+    return url.toString().substring(0, url.toString().length()
+                                       - (url.getQuery() == null ? 0
+                                           : url.getQuery().length())
+                                       - (url.getRef() == null ? 0
+                                           : url.getRef().length() + 1))
+           + " " + (url.getRef() == null ? "" : url.getRef());
```

Comment:
this piece of code is hard to understand!  
My first suggestion would be to see this method can be implemented simply by calling  `StringUtils.remove(url.toString(), url.getQuery());` if that does not work then a few things could make it more readable:
* introduce local variables for `url.getQuery()` and `url.getRef()`
* `s == null ? 0 : s.length()` is code repeated twice for `url.getQuery()` and `url.getRef()` and should be replaced by `StringUtils.length`

Reply:
Evidently you are right. How about this?

```java
  public void assertIsEqualToWithSortedQueryParameters(AssertionInfo info, URL actual, URL expected) {
    assertNotNull(info, actual);
    String expectedQuery = expected.getQuery() == null ? "" : expected.getQuery();
    String actualQuery = actual.getQuery() == null ? "" : actual.getQuery();
    String expectedNonQueryPart = expected.toString().replace(expectedQuery, " ");
    String actualNonQueryPart = actual.toString().replace(actualQuery, " ");

    if (!expectedNonQueryPart.equals(actualNonQueryPart))
      throw failures.failure(info, shouldBeEqualToWithSortedQueryParameters(actual, expected));

    String[] expectedQueryParams = extractQueryParams(expected);
    String[] actualQueryParams = extractQueryParams(actual);
    Arrays.sort(expectedQueryParams);
    Arrays.sort(actualQueryParams);
    if (!Objects.deepEquals(expectedQueryParams, actualQueryParams))
      throw failures.failure(info, shouldBeEqualToWithSortedQueryParameters(actual, expected));
  }
```

Reply:
All my test cases are passed by this code snippet.